File: scripts/lib/cli_framework.py
#!/usr/bin/env python3
"""
Shared CLI Framework for MokoStandards
Version: 03.02.00
Copyright (C) 2026 Moko Consulting LLC

Provides consistent CLI interface for all MokoStandards scripts:
- Common argument parsing
- Standard help formatting
- Consistent error handling
- Integrated logging setup
- Enterprise library integration

Usage:
    from cli_framework import CLIApp
    
    class MyScript(CLIApp):
        def setup_arguments(self):
            self.parser.add_argument('--input', help='Input file')
        
        def run(self):
            print(f"Processing: {self.args.input}")
    
    if __name__ == '__main__':
        MyScript().execute()
"""

import argparse
import logging
import sys
from abc import ABC, abstractmethod
from datetime import datetime
from pathlib import Path
from typing import Any, Dict, List, Optional

VERSION = "03.02.00"


class CLIApp(ABC):
    """Base class for CLI applications with common functionality."""
    
    def __init__(self, name: str = None, description: str = None, version: str = VERSION):
        """Initialize CLI application.
        
        Args:
            name: Application name (defaults to class name)
            description: Application description
            version: Application version
        """
        self.name = name or self.__class__.__name__
        self.description = description or f"{self.name} - MokoStandards CLI Tool"
        self.version = version
        
        # Create argument parser
        self.parser = argparse.ArgumentParser(
            prog=self.name,
            description=self.description,
            formatter_class=argparse.RawDescriptionHelpFormatter
        )
        
        # Add common arguments
        self._add_common_arguments()
        
        # Subclass adds its own arguments
        self.setup_arguments()
        
        # Parse args (will be set by execute())
        self.args = None
        self.logger = None
        
        # Enterprise library instances (optional)
        self.audit_logger = None
        self.metrics = None
        self.config = None
        
    def _add_common_arguments(self):
        """Add common arguments available to all CLI apps."""
        self.parser.add_argument(
            '--version',
            action='version',
            version=f'{self.name} v{self.version}'
        )
        self.parser.add_argument(
            '--verbose', '-v',
            action='store_true',
            help='Enable verbose output'
        )
        self.parser.add_argument(
            '--quiet', '-q',
            action='store_true',
            help='Suppress non-error output'
        )
        self.parser.add_argument(
            '--dry-run',
            action='store_true',
            help='Perform dry run without making changes'
        )
        self.parser.add_argument(
            '--log-file',
            type=str,
            help='Write logs to file'
        )
        self.parser.add_argument(
            '--json',
            action='store_true',
            help='Output results in JSON format'
        )
        self.parser.add_argument(
            '--audit',
            action='store_true',
            help='Enable audit logging'
        )
        self.parser.add_argument(
            '--metrics',
            action='store_true',
            help='Collect and display metrics'
        )
    
    @abstractmethod
    def setup_arguments(self):
        """Setup script-specific arguments.
        
        Override this method to add custom arguments to self.parser
        """
        pass
    
    @abstractmethod
    def run(self) -> int:
        """Main execution logic.
        
        Override this method with your script's logic.
        Access parsed arguments via self.args
        
        Returns:
            Exit code (0 for success, non-zero for error)
        """
        pass
    
    def setup_logging(self):
        """Setup logging based on arguments."""
        if self.args.quiet:
            level = logging.ERROR
        elif self.args.verbose:
            level = logging.DEBUG
        else:
            level = logging.INFO
        
        log_format = '%(asctime)s - %(name)s - %(levelname)s - %(message)s'
        
        if self.args.log_file:
            logging.basicConfig(
                level=level,
                format=log_format,
                filename=self.args.log_file,
                filemode='a'
            )
        else:
            logging.basicConfig(
                level=level,
                format=log_format
            )
        
        self.logger = logging.getLogger(self.name)
    
    def setup_enterprise_features(self):
        """Setup enterprise features if enabled."""
        # Setup audit logging
        if self.args.audit:
            try:
                from enterprise_audit import AuditLogger
                self.audit_logger = AuditLogger(service=self.name)
                self.logger.info("Audit logging enabled")
            except ImportError:
                self.logger.warning("Audit logging requested but enterprise_audit not available")
        
        # Setup metrics collection
        if self.args.metrics:
            try:
                from metrics_collector import MetricsCollector
                self.metrics = MetricsCollector(service_name=self.name)
                self.logger.info("Metrics collection enabled")
            except ImportError:
                self.logger.warning("Metrics requested but metrics_collector not available")
        
        # Load configuration
        try:
            from config_manager import Config
            self.config = Config.load()
            self.logger.debug("Configuration loaded")
        except ImportError:
            self.logger.debug("Config manager not available")
    
    def execute(self, args: Optional[List[str]] = None) -> int:
        """Execute the CLI application.
        
        Args:
            args: Command line arguments (defaults to sys.argv[1:])
            
        Returns:
            Exit code
        """
        try:
            # Parse arguments
            self.args = self.parser.parse_args(args)
            
            # Setup logging
            self.setup_logging()
            
            # Setup enterprise features
            self.setup_enterprise_features()
            
            # Log start
            self.logger.info(f"Starting {self.name} v{self.version}")
            if self.args.dry_run:
                self.logger.info("DRY RUN MODE - No changes will be made")
            
            # Run main logic
            if self.metrics:
                with self.metrics.timer('main_execution'):
                    exit_code = self.run()
            else:
                exit_code = self.run()
            
            # Log completion
            self.logger.info(f"Completed {self.name} with exit code {exit_code}")
            
            # Print metrics if enabled
            if self.metrics:
                self.metrics.print_summary()
            
            return exit_code
            
        except KeyboardInterrupt:
            self.logger.warning("Interrupted by user")
            return 130
        except Exception as e:
            self.logger.error(f"Unhandled exception: {e}", exc_info=True)
            return 1
    
    def print_result(self, result: Any, json_output: bool = None):
        """Print result in appropriate format.
        
        Args:
            result: Result to print
            json_output: Use JSON format (defaults to self.args.json)
        """
        if json_output is None:
            json_output = self.args.json if self.args else False
        
        if json_output:
            import json
            print(json.dumps(result, indent=2))
        else:
            print(result)
    
    def confirm(self, message: str, default: bool = False) -> bool:
        """Ask for user confirmation.
        
        Args:
            message: Confirmation message
            default: Default response
            
        Returns:
            True if user confirms
        """
        if self.args.dry_run:
            self.logger.info(f"[DRY RUN] Would ask: {message}")
            return False
        
        suffix = " [Y/n]" if default else " [y/N]"
        response = input(message + suffix + ": ").strip().lower()
        
        if not response:
            return default
        
        return response in ['y', 'yes']


class ValidationCLI(CLIApp):
    """CLI for validation operations."""
    
    def setup_arguments(self):
        """Setup validation-specific arguments."""
        self.parser.add_argument(
            '--check',
            choices=['all', 'paths', 'markdown', 'licenses', 'workflows', 'security'],
            default='all',
            help='Type of validation to perform'
        )
        self.parser.add_argument(
            '--dir',
            type=str,
            default='.',
            help='Directory to validate'
        )
    
    def run(self) -> int:
        """Run validation checks."""
        self.logger.info(f"Running validation: {self.args.check}")
        
        # Use unified validation framework
        try:
            from unified_validation import UnifiedValidator, PathValidatorPlugin
            
            validator = UnifiedValidator()
            
            if self.args.check in ['all', 'paths']:
                validator.add_plugin(PathValidatorPlugin())
            
            # Create context
            context = {
                'paths': [self.args.dir],
                'scan_dir': self.args.dir
            }
            
            # Run validation
            results = validator.validate_all(context)
            
            # Print results
            if not self.args.json:
                validator.print_summary()
            else:
                result_data = [
                    {
                        'plugin': r.plugin_name,
                        'passed': r.passed,
                        'message': r.message,
                        'details': r.details
                    }
                    for r in results
                ]
                self.print_result(result_data)
            
            return 0 if validator.all_passed() else 1
            
        except ImportError as e:
            self.logger.error(f"Validation framework not available: {e}")
            return 1


class MaintenanceCLI(CLIApp):
    """CLI for maintenance operations."""
    
    def setup_arguments(self):
        """Setup maintenance-specific arguments."""
        self.parser.add_argument(
            '--operation',
            choices=['cleanup', 'update', 'status'],
            required=True,
            help='Maintenance operation to perform'
        )
        self.parser.add_argument(
            '--target',
            type=str,
            help='Target for maintenance operation'
        )
    
    def run(self) -> int:
        """Run maintenance operation."""
        self.logger.info(f"Running maintenance: {self.args.operation}")
        
        if self.args.operation == 'cleanup':
            self.logger.info("Performing cleanup...")
            # Cleanup logic here
            return 0
        elif self.args.operation == 'status':
            self.logger.info("Checking status...")
            # Status logic here
            return 0
        
        return 0


# Example usage and testing
if __name__ == "__main__":
    print(f"CLI Framework v{VERSION}")
    print("=" * 50)
    
    # Test 1: Basic CLI app
    print("\n1. Testing basic CLI app...")
    
    class TestApp(CLIApp):
        def setup_arguments(self):
            self.parser.add_argument('--name', default='World', help='Name to greet')
        
        def run(self):
            self.logger.info(f"Hello, {self.args.name}!")
            return 0
    
    app = TestApp(name="test_app", description="Test application")
    exit_code = app.execute(['--name', 'MokoStandards', '--verbose'])
    print(f"   ✓ Exit code: {exit_code}")
    
    # Test 2: Validation CLI
    print("\n2. Testing validation CLI...")
    val_app = ValidationCLI(name="validation", description="Validation tool")
    exit_code = val_app.execute(['--check', 'all', '--dir', '.', '--quiet'])
    print(f"   ✓ Validation exit code: {exit_code}")
    
    # Test 3: With metrics
    print("\n3. Testing with metrics...")
    app2 = TestApp()
    exit_code = app2.execute(['--metrics', '--audit'])
    print(f"   ✓ Exit code with metrics: {exit_code}")
    
    print("\n✓ All tests passed!")

File: src/Enterprise/CliFramework.php
<?php

declare(strict_types=1);

/**
 * Shared CLI Framework for MokoStandards
 *
 * Provides consistent CLI interface for all MokoStandards scripts:
 * - Common argument parsing with getopt
 * - Standard help formatting
 * - Consistent error handling
 * - Integrated logging setup
 * - Enterprise library integration
 * - Command routing and subcommands
 * - Dry-run mode support
 * - JSON output formatting
 *
 * Example usage:
 * ```php
 * class MyScript extends CLIApp {
 *     protected function setupArguments(): array {
 *         return [
 *             'input:' => 'Input file path',
 *             'output:' => 'Output file path'
 *         ];
 *     }
 *     
 *     protected function run(): int {
 *         $input = $this->getOption('input');
 *         echo "Processing: {$input}\n";
 *         return 0;
 *     }
 * }
 * 
 * $app = new MyScript('my_script', 'My Script Description');
 * exit($app->execute());
 * ```
 *
 * Copyright (C) 2026 Moko Consulting <hello@mokoconsulting.tech>
 *
 * SPDX-License-Identifier: GPL-3.0-or-later
 *
 * @package MokoStandards\Enterprise
 * @version 04.00.00
 * @author MokoStandards Team
 * @license GPL-3.0-or-later
 */

namespace MokoStandards\Enterprise;

use DateTime;
use DateTimeZone;
use Exception;

/**
 * Base class for CLI applications with common functionality
 */
abstract class CLIApp
{
    private const VERSION = '04.00.00';

    protected string $name;
    protected string $description;
    protected string $version;
    protected array $options = [];
    protected array $arguments = [];
    protected bool $verbose = false;
    protected bool $quiet = false;
    protected bool $dryRun = false;
    protected bool $jsonOutput = false;
    
    // Enterprise features
    protected ?MetricsCollector $metrics = null;
    protected ?object $auditLogger = null;

    public function __construct(string $name, string $description = '', string $version = self::VERSION)
    {
        $this->name = $name;
        $this->description = $description ?: "{$name} - MokoStandards CLI Tool";
        $this->version = $version;
    }

    /**
     * Setup script-specific arguments
     * 
     * Return an associative array where keys are option specs and values are descriptions.
     * Option spec format: 'name:' for required value, 'name::' for optional value, 'name' for flag
     *
     * @return array<string, string> Option specifications and descriptions
     */
    abstract protected function setupArguments(): array;

    /**
     * Main execution logic
     *
     * @return int Exit code (0 for success, non-zero for error)
     */
    abstract protected function run(): int;

    /**
     * Get common CLI options
     *
     * @return array<string, string> Common options
     */
    protected function getCommonOptions(): array
    {
        return [
            'version' => 'Display version information',
            'verbose' => 'Enable verbose output',
            'v' => 'Alias for --verbose',
            'quiet' => 'Suppress non-error output',
            'q' => 'Alias for --quiet',
            'dry-run' => 'Perform dry run without making changes',
            'json' => 'Output results in JSON format',
            'metrics' => 'Collect and display metrics',
            'help' => 'Display this help message',
            'h' => 'Alias for --help',
        ];
    }

    /**
     * Parse command line arguments
     */
    protected function parseArguments(): void
    {
        $shortOpts = 'vqh';
        $longOpts = [
            'version',
            'verbose',
            'quiet',
            'dry-run',
            'json',
            'metrics',
            'help',
        ];

        // Add custom options
        $customOpts = $this->setupArguments();
        foreach (array_keys($customOpts) as $opt) {
            if (str_ends_with($opt, '::')) {
                $longOpts[] = rtrim($opt, ':') . '::';
            } elseif (str_ends_with($opt, ':')) {
                $longOpts[] = rtrim($opt, ':') . ':';
            } else {
                $longOpts[] = $opt;
            }
        }

        $this->options = getopt($shortOpts, $longOpts);
        $this->arguments = array_slice($_SERVER['argv'], 1);

        // Handle common flags
        if (isset($this->options['version'])) {
            echo "{$this->name} v{$this->version}\n";
            exit(0);
        }

        if (isset($this->options['help']) || isset($this->options['h'])) {
            $this->printHelp();
            exit(0);
        }

        $this->verbose = isset($this->options['verbose']) || isset($this->options['v']);
        $this->quiet = isset($this->options['quiet']) || isset($this->options['q']);
        $this->dryRun = isset($this->options['dry-run']);
        $this->jsonOutput = isset($this->options['json']);
    }

    /**
     * Print help message
     */
    protected function printHelp(): void
    {
        echo "{$this->description}\n\n";
        echo "Usage: {$this->name} [options]\n\n";
        echo "Options:\n";

        $allOpts = array_merge($this->getCommonOptions(), $this->setupArguments());
        
        foreach ($allOpts as $opt => $desc) {
            $optName = rtrim($opt, ':');
            $hasValue = str_ends_with($opt, ':');
            $optDisplay = strlen($optName) === 1 ? "-{$optName}" : "--{$optName}";
            
            if ($hasValue) {
                $optDisplay .= ' <value>';
            }
            
            echo sprintf("  %-30s %s\n", $optDisplay, $desc);
        }
        
        echo "\n";
    }

    /**
     * Setup logging
     */
    protected function setupLogging(): void
    {
        if ($this->quiet) {
            error_reporting(E_ERROR);
        } elseif ($this->verbose) {
            error_reporting(E_ALL);
        }
    }

    /**
     * Setup enterprise features
     */
    protected function setupEnterpriseFeatures(): void
    {
        if (isset($this->options['metrics'])) {
            try {
                $this->metrics = new MetricsCollector($this->name);
                $this->log("Metrics collection enabled", 'INFO');
            } catch (Exception $e) {
                $this->log("Metrics collection unavailable: " . $e->getMessage(), 'WARNING');
            }
        }
    }

    /**
     * Execute the CLI application
     *
     * @return int Exit code
     */
    public function execute(): int
    {
        try {
            $this->parseArguments();
            $this->setupLogging();
            $this->setupEnterpriseFeatures();

            $this->log("Starting {$this->name} v{$this->version}", 'INFO');
            
            if ($this->dryRun) {
                $this->log("DRY RUN MODE - No changes will be made", 'INFO');
            }

            $startTime = microtime(true);
            
            if ($this->metrics !== null) {
                $timer = $this->metrics->startTimer('main_execution');
                $exitCode = $this->run();
                $timer->stop($exitCode === 0);
            } else {
                $exitCode = $this->run();
            }

            $duration = microtime(true) - $startTime;
            $this->log(sprintf("Completed {$this->name} with exit code %d (%.2fs)", $exitCode, $duration), 'INFO');

            if ($this->metrics !== null && !$this->quiet) {
                $this->metrics->printSummary();
            }

            return $exitCode;
        } catch (Exception $e) {
            $this->log("Unhandled exception: " . $e->getMessage(), 'ERROR');
            if ($this->verbose) {
                $this->log($e->getTraceAsString(), 'ERROR');
            }
            return 1;
        }
    }

    /**
     * Get option value
     *
     * @param string $name Option name
     * @param mixed $default Default value if not set
     * @return mixed Option value
     */
    protected function getOption(string $name, $default = null)
    {
        return $this->options[$name] ?? $default;
    }

    /**
     * Check if option is set
     *
     * @param string $name Option name
     * @return bool True if option is set
     */
    protected function hasOption(string $name): bool
    {
        return isset($this->options[$name]);
    }

    /**
     * Log a message
     *
     * @param string $message Message to log
     * @param string $level Log level (INFO, WARNING, ERROR)
     */
    protected function log(string $message, string $level = 'INFO'): void
    {
        if ($this->quiet && $level !== 'ERROR') {
            return;
        }

        if (!$this->verbose && $level === 'DEBUG') {
            return;
        }

        $timestamp = (new DateTime('now', new DateTimeZone('UTC')))->format('Y-m-d H:i:s');
        $formatted = "[{$timestamp}] {$level}: {$message}\n";
        
        if ($level === 'ERROR') {
            fwrite(STDERR, $formatted);
        } else {
            echo $formatted;
        }
    }

    /**
     * Print result in appropriate format
     *
     * @param mixed $result Result to print
     */
    protected function printResult($result): void
    {
        if ($this->jsonOutput) {
            echo json_encode($result, JSON_PRETTY_PRINT) . "\n";
        } else {
            if (is_array($result)) {
                print_r($result);
            } else {
                echo $result . "\n";
            }
        }
    }

    /**
     * Ask for user confirmation
     *
     * @param string $message Confirmation message
     * @param bool $default Default response
     * @return bool True if user confirms
     */
    protected function confirm(string $message, bool $default = false): bool
    {
        if ($this->dryRun) {
            $this->log("[DRY RUN] Would ask: {$message}", 'INFO');
            return false;
        }

        $suffix = $default ? ' [Y/n]' : ' [y/N]';
        echo $message . $suffix . ': ';
        
        $handle = fopen('php://stdin', 'r');
        $response = trim(fgets($handle));
        fclose($handle);

        if (empty($response)) {
            return $default;
        }

        return in_array(strtolower($response), ['y', 'yes'], true);
    }

    /**
     * Print colored output (if terminal supports it)
     *
     * @param string $text Text to print
     * @param string $color Color name (red, green, yellow, blue)
     */
    protected function printColored(string $text, string $color): void
    {
        $colors = [
            'red' => "\033[31m",
            'green' => "\033[32m",
            'yellow' => "\033[33m",
            'blue' => "\033[34m",
            'reset' => "\033[0m",
        ];

        if (isset($colors[$color]) && posix_isatty(STDOUT)) {
            echo $colors[$color] . $text . $colors['reset'];
        } else {
            echo $text;
        }
    }

    public function getVersion(): string
    {
        return $this->version;
    }
}

/**
 * CLI for validation operations
 */
class ValidationCLI extends CLIApp
{
    protected function setupArguments(): array
    {
        return [
            'check:' => 'Type of validation (all, paths, markdown, licenses, workflows, security)',
            'dir:' => 'Directory to validate (default: current directory)',
        ];
    }

    protected function run(): int
    {
        $check = $this->getOption('check', 'all');
        $dir = $this->getOption('dir', '.');

        $this->log("Running validation: {$check}", 'INFO');

        try {
            $validator = new UnifiedValidator();

            if (in_array($check, ['all', 'paths'], true)) {
                $validator->addPlugin(new PathValidatorPlugin());
            }
            if (in_array($check, ['all', 'markdown'], true)) {
                $validator->addPlugin(new MarkdownValidatorPlugin());
            }

            $context = [
                'paths' => [$dir],
                'scan_dir' => $dir,
            ];

            $results = $validator->validateAll($context);

            if (!$this->jsonOutput) {
                $validator->printSummary();
            } else {
                $resultData = array_map(function ($r) {
                    return [
                        'plugin' => $r->pluginName,
                        'passed' => $r->passed,
                        'message' => $r->message,
                        'details' => $r->details,
                    ];
                }, $results);
                $this->printResult($resultData);
            }

            return $validator->allPassed() ? 0 : 1;
        } catch (Exception $e) {
            $this->log("Validation error: " . $e->getMessage(), 'ERROR');
            return 1;
        }
    }
}
